modal/_utils/deprecation.py

```diff
@@ -0,0 +1,75 @@
+# Copyright Modal Labs 2024
+import functools
+import sys
+import warnings
+from datetime import date
+from typing import Callable, TypeVar
+
+from typing_extensions import ParamSpec  # Needed for Python 3.9
+
+from ..exception import DeprecationError, PendingDeprecationError
+
+# TODO(erikbern): we have something similready in function_utils.py
+_INTERNAL_MODULES = ["modal", "synchronicity"]
+
+
+def _is_internal_frame(frame):
+    module = frame.f_globals["__name__"].split(".")[0]
+    return module in _INTERNAL_MODULES
+
+
+def deprecation_error(deprecated_on: tuple[int, int, int], msg: str):
+    raise DeprecationError(f"Deprecated on {date(*deprecated_on)}: {msg}")
+
+
+def deprecation_warning(
+    deprecated_on: tuple[int, int, int], msg: str, *, pending: bool = False, show_source: bool = True
+) -> None:
+    """Utility for getting the proper stack entry.
+
+    See the implementation of the built-in [warnings.warn](https://docs.python.org/3/library/warnings.html#available-functions).
+    """
+    filename, lineno = "<unknown>", 0
+    if show_source:
+        # Find the last non-Modal line that triggered the warning
+        try:
+            frame = sys._getframe()
+            while frame is not None and _is_internal_frame(frame):
+                frame = frame.f_back
+            filename = frame.f_code.co_filename
+            lineno = frame.f_lineno
+        except ValueError:
+            # Use the defaults from above
+            pass
+
+    warning_cls: type = PendingDeprecationError if pending else DeprecationError
+
+    # This is a lower-level function that warnings.warn uses
+    warnings.warn_explicit(f"{date(*deprecated_on)}: {msg}", warning_cls, filename, lineno)
+
+
+P = ParamSpec("P")
+R = TypeVar("R")
+
+
+def renamed_parameter(
```

Comment:
would it be simpler if this isn't a decorator and we just have code like this
```python
def f(
    new_parameter: int = 42,
    old_parameter: Optional[int] = None
):
    if old_parameter is not None:
        deprecation_warning(...)
        new_parameter = old_parameter
```

Reply:
Simpler in the sense of avoiding having to fight with whatever synchronicity is doing but it will be a lot more clutter in the code and produce signatures that are less clear about the right way going forward. I would call that more complex?

Also some of the parameters that we intent to rename have `None` as a valid argument so we'd need to introduce a "default" sentinel ... again doable just more headache. The problem is that synchronicity is too clever.

Reply:
Probably more lines in total but I think the maintenance cost of those lines is very low – the behavior is self-explanatory and they don't interfere with anything else.

Two problems with decorator is (a) typing (b) longer tracebacks. Those are issues even without synchronicity. I do agree that synchronicity blows things up though!

Not a hill I would die on but I would have personally just used "dumb" if-statements for these deprecations. Speculative but it also helps if we want to do any sort of value transformation.